controllers/microk8sconfig_controller.go

```diff
@@ -335,6 +329,91 @@ func (r *MicroK8sConfigReconciler) handleJoiningControlPlaneNode(ctx context.Con
 	return ctrl.Result{}, nil
 }
 
+func (r *MicroK8sConfigReconciler) handleJoiningWorkerNode(ctx context.Context, scope *Scope) (_ ctrl.Result, reterr error) {
+	// if it's a control plane machine, requeue
+	if scope.ConfigOwner.IsControlPlaneMachine() {
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	machine := &clusterv1.Machine{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(scope.ConfigOwner.Object, machine); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "cannot convert %s to Machine", scope.ConfigOwner.GetKind())
+	}
+
+	// acquire the init lock so that only only one machine joins each time
+	if !r.MicroK8sInitLock.Lock(ctx, scope.Cluster, machine) {
+		scope.Info("A node is already being handled, requeueing until cluster can be extended with this node")
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	defer func() {
+		if reterr != nil {
+			if !r.MicroK8sInitLock.Unlock(ctx, scope.Cluster) {
+				reterr = kerrors.NewAggregate([]error{reterr, errors.New("failed to unlock the init lock")})
+			}
+		}
+	}()
+
+	scope.Info("Creating BootstrapData for the joining worker")
+
+	microk8sConfig := &bootstrapclusterxk8siov1beta1.MicroK8sConfig{}
+	if err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: machine.Spec.Bootstrap.ConfigRef.Namespace,
+		Name:      machine.Spec.Bootstrap.ConfigRef.Name,
+	}, microk8sConfig); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// TODO: Make this port configurable. Make sure you address the same issue in the controller provider
+	portOfNodeToConnectTo := "2379"
+	ipOfNodeToConnectTo := ""
+
+	nodes, err := r.getControlPlaneMachinesForCluster(ctx, util.ObjectKey(scope.Cluster))
+	if err != nil {
+		scope.Error(err, "Lookup control plane nodes")
+		return ctrl.Result{}, err
+	}
+
+	for _, node := range nodes {
+		if ipOfNodeToConnectTo != "" {
+			break
+		}
+		if node.Spec.ProviderID != nil && node.Status.Phase == "Running" {
```

Comment:
Shouldn't only the node that ran `microk8s add-node` be used?

Reply:
We inject the same join token to all nodes so it does not matter what control plane node we will contact during the join command.